### PR TITLE
[extractor/ceskatelevize] Back-port extractor from yt-dlp, etc

### LINKED ISSUE #29415: [ceskatelevize] youtube_dl.utils.ExtractorError: No video formats found
## Checklist
- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2021.06.06**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
usansky$ youtube-dl --verbose https://www.ceskatelevize.cz/porady/10597292024-filmovy-dobrodruh-karel-zeman/31429434015/
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--verbose', 'https://www.ceskatelevize.cz/porady/10597292024-filmovy-dobrodruh-karel-zeman/31429434015/']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] youtube-dl version 2021.06.06
[debug] Python version 3.8.10 (CPython) - Linux-5.12.12-1-default-x86_64-with-glibc2.2.5
[debug] exe versions: ffmpeg 4.4, ffprobe 4.4
[debug] Proxy map: {}
[CeskaTelevizePorady] 31429434015: Downloading webpage
[CeskaTelevize] iFramePlayer.php: Downloading webpage
[CeskaTelevize] iFramePlayer.php: Downloading JSON metadata
[CeskaTelevize] iFramePlayer.php: Downloading JSON metadata
[CeskaTelevize] iFramePlayer.php: Downloading JSON metadata
[CeskaTelevize] iFramePlayer.php: Downloading JSON metadata
ERROR: No video formats found; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 815, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 836, in __extract_info
    ie_result = ie.extract(url)
  File "/usr/bin/youtube-dl/youtube_dl/extractor/common.py", line 534, in extract
    ie_result = self._real_extract(url)
  File "/usr/bin/youtube-dl/youtube_dl/extractor/ceskatelevize.py", line 201, in _real_extract
    self._sort_formats(e['formats'])
  File "/usr/bin/youtube-dl/youtube_dl/extractor/common.py", line 1374, in _sort_formats
    raise ExtractorError('No video formats found')
youtube_dl.utils.ExtractorError: No video formats found; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.

~@kusansky$
```

## Description
Yes, this video works in Firefox on my desktop, but it just doesn't download at all with youtube-dl.

### LINKED ISSUE #30151: Broken support of ČT iVysílání
## Checklist

- [ ] I'm reporting a broken site support issue
- [x] I've verified that I'm running youtube-dl version **2021.06.06**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [ ] I've searched the bugtracker for similar bug reports including closed ones
- [ ] I've read bugs section in FAQ

## Verbose log

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--verbose', 'https://www.ceskatelevize.cz/ivysilani/1097147804-az-kviz/321291310010133/']
[debug] Encodings: locale cp1250, fs mbcs, out cp852, pref cp1250
[debug] youtube-dl version 2021.06.06
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.22000
[debug] exe versions: ffmpeg 2021-10-21-git-2aa343bb6f-full_build-www.gyan.dev
[debug] Proxy map: {}
[CeskaTelevize] 321291310010133: Downloading webpage
[CeskaTelevize] 321291310010133: Downloading JSON metadata
Traceback (most recent call last):
  File "__main__.py", line 19, in <module>
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\__init__.py", line 475, in main
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\__init__.py", line 465, in _real_main
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\YoutubeDL.py", line 2069, in download
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\YoutubeDL.py", line 808, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\YoutubeDL.py", line 815, in wrapper
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\YoutubeDL.py", line 836, in __extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\extractor\common.py", line 534, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\extractor\ceskatelevize.py", line 130, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\utils.py", line 2158, in sanitized_Request
  File "C:\Python\Python34\lib\urllib\request.py", line 267, in __init__
  File "C:\Python\Python34\lib\urllib\request.py", line 293, in full_url
  File "C:\Python\Python34\lib\urllib\request.py", line 322, in _parse
ValueError: unknown url type: 'Error'
```

## Description

Support of Czech ČT iVysílání VOD service was broken! It was still working on Friday 22 October. Invited to resolve this problem are especially contributors from Czech Republic.

### LINKED ISSUE #30166: problem parsing site ceskatelevize.cz while trying to downloading video
## Checklist

- [x ] I'm reporting a broken site support
- [ x] I've verified that I'm running youtube-dl version **2021.06.06**
- [ x] I've checked that all provided URLs are alive and playable in a browser
- [ x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [ x] I've searched the bugtracker for similar issues including closed ones

## Verbose log
```
c:\YoutubeDL>youtube-dl.exe --no-check-certificate --no-mtime -F https://www.ceskatelevize.cz/ivysilani/19796-pumpari-od-zlate-podkovy/29238360846/ -v
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--no-check-certificate', '--no-mtime', '-F', 'https://www.ceskatelevize.cz/ivysilani/19796-pumpari-od-zlate-podkovy/29238360846/', '-v']
[debug] Encodings: locale cp1252, fs mbcs, out cp437, pref cp1252
[debug] youtube-dl version 2021.06.06
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.19041
[debug] exe versions: ffmpeg N-90920-ge07b1913fc, ffprobe N-90920-ge07b1913fc
[debug] Proxy map: {}
[CeskaTelevize] 29238360846: Downloading webpage
[CeskaTelevize] 29238360846: Downloading JSON metadata
Traceback (most recent call last):
  File "__main__.py", line 19, in <module>
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\__init__.py", line 475, in main
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\__init__.py", line 465, in _real_main
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\YoutubeDL.py", line 2069, in download
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\YoutubeDL.py", line 808, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\YoutubeDL.py", line 815, in wrapper
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\YoutubeDL.py", line 836, in __extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\extractor\common.py", line 534, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\extractor\ceskatelevize.py", line 130, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\utils.py", line 2158, in sanitized_Request
  File "C:\Python\Python34\lib\urllib\request.py", line 267, in __init__
  File "C:\Python\Python34\lib\urllib\request.py", line 293, in full_url
  File "C:\Python\Python34\lib\urllib\request.py", line 322, in _parse
ValueError: unknown url type: 'Error'
```

## Description

Download form ceskatelevize.cz is not working - parsing error

### LINKED ISSUE #30172: 【CeskaTelevize.CZ】 Download NO Longer Possible
## Checklist

- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2021.06.06**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
C:\Users\DELL\video\TEST>youtube-dl --all-subs --convert-subs srt --retries 20000 -f hls-main-6272 -v https://www.ceskatelevize.cz/ivysilani/10111864738-hercule-poirot/211381467540018-tajemstvi-levneho-bytu
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--all-subs', '--convert-subs', 'srt', '--retries', '20000', '-f', 'hls-main-6272', '-v', 'https://www.ceskatelevize.cz/ivysilani/10111864738-hercule-poirot/211381467540018-tajemstvi-levneho-bytu']
[debug] Encodings: locale cp936, fs mbcs, out cp936, pref cp936
[debug] youtube-dl version 2021.06.06
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.18362
[debug] exe versions: ffmpeg N-89508-g1f1207145a, ffprobe N-89508-g1f1207145a
[debug] Proxy map: {'https': 'https://192.168.49.1:8282', 'ftp': 'ftp://192.168.49.1:8282', 'http': 'http://192.168.49.1:8282'}
[CeskaTelevize] 211381467540018-tajemstvi-levneho-bytu: Downloading webpage
[CeskaTelevize] 211381467540018-tajemstvi-levneho-bytu: Downloading JSON metadata
Traceback (most recent call last):
  File "__main__.py", line 19, in <module>
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\__init__.py", line 475, in main
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\__init__.py", line 465, in _real_main
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\YoutubeDL.py", line 2069, in download
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\YoutubeDL.py", line 808, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\YoutubeDL.py", line 815, in wrapper
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\YoutubeDL.py", line 836, in __extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\extractor\common.py", line 534, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\extractor\ceskatelevize.py", line 130, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpkqxnwl31\build\youtube_dl\utils.py", line 2158, in sanitized_Request
  File "C:\Python\Python34\lib\urllib\request.py", line 267, in __init__
  File "C:\Python\Python34\lib\urllib\request.py", line 293, in full_url
  File "C:\Python\Python34\lib\urllib\request.py", line 322, in _parse
ValueError: unknown url type: 'Error'

C:\Users\DELL\video\TEST>
```

## Description

I've tried https://www.ceskatelevize.cz/ivysilani/12878558019-desatero-hrichu/220381485260005-desatero-hrichu-ii as well & the download doesn't work either. Chances are that they have updated the video format?

Can we please have this looked into ASAP and make sure the solution is available for **Windows 10**, i.e. can we have an updated ALL-IN-ONE "**youtube-dl.exe**" file for us **Windows 10** users?

Thank you
NiesmialyGosc

### LINKED ISSUE #30406: CeskaTelevize does not work anymore
## Checklist

- [ x] I'm reporting a broken site support
- [ x] I've verified that I'm running youtube-dl version **2021.12.17**
- [ x] I've checked that all provided URLs are alive and playable in a browser
- [ x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [ x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['https://www.ceskatelevize.cz/porady/12833268908-staroveke-rise/219382567680006/', '--verbose']
[debug] Encodings: locale cp1250, fs mbcs, out cp852, pref cp1250
[debug] youtube-dl version 2021.12.17
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.19041
[debug] exe versions: ffmpeg 3.0.8
[debug] Proxy map: {}
[CeskaTelevizePorady] 219382567680006: Downloading webpage
ERROR: Unable to extract iframe player url; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\YoutubeDL.py", line 815, in wrapper
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\YoutubeDL.py", line 836, in __extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\extractor\common.py", line 534, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\extractor\ceskatelevize.py", line 285, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\extractor\common.py", line 1012, in _search_regex
youtube_dl.utils.RegexNotFoundError: Unable to extract iframe player url; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.


```

## Description

After redesign and new version of iVysilani of CeskaTelevize, downloading of videos does not work with "ERROR: Unable to extract iframe player url". For example URl  https://www.ceskatelevize.cz/porady/12833268908-staroveke-rise/219382567680006/

Thanks for updating of  youtube-dl

### LINKED ISSUE #30696: New Czech TV web-site does not work
## Checklist

- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2021.12.17**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--no-check-certificate', '--verbose', '--no-overwrites', '--restrict-filenames', '--output', 'C:\\Temp\\%(title)s_%(autonumber)s.%(ext)s', '--batch-file', 'C:\\Program Files (x86)\\YouTube-DL\\youtube-dl.lst']
[debug] Batch file urls: ['www.ceskatelevize.cz/ivysilani/embed/iFramePlayer.php?hash=18f5f760865c2ea87d0952ff185d2983a060cc58']
[debug] Encodings: locale cp1250, fs mbcs, out cp852, pref cp1250
[debug] youtube-dl version 2021.12.17
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.19041
[debug] exe versions: ffmpeg 4.3.2-2021-02-20-full_build-www.gyan.dev, ffprobe 4.3.2-2021-02-20-full_build-www.gyan.dev
[debug] Proxy map: {}
WARNING: The url doesn't specify the protocol, trying with http
[CeskaTelevize] iFramePlayer.php: Downloading webpage
ERROR: Unable to extract type; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\YoutubeDL.py", line 815, in wrapper
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\YoutubeDL.py", line 836, in __extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\extractor\common.py", line 534, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\extractor\ceskatelevize.py", line 94, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\extractor\common.py", line 1021, in _html_search_regex
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\extractor\common.py", line 1012, in _search_regex
youtube_dl.utils.RegexNotFoundError: Unable to extract type; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

## Description

New formats of Czech TV video-files does not work...
Example 1: 
https://www.ceskatelevize.cz/porady/13875183193-zappa/22138775999/
Example 2:
https://www.ceskatelevize.cz/ivysilani/embed/iFramePlayer.php?hash=18f5f760865c2ea87d0952ff185d2983a060cc58&origin=iVysilani&autoStart=true&width=100%25&IDEC=22138775999&start=0

### PR DESCRIPTION
**Pull request by dirkf**
## Please follow the guide below

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/), except for code acquired from yt-dlp for which either this or the below was previously asserted
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This PR back-ports the CeskaTelevize yt-dlp extractor, removing CeskaTelevizePoradyIE. An interim `traverse_obj()` shim is used. The existing tests have been fixed and/or new ones added as needed.

The following additional changes were made:
* follow redirect URL
* support `liveBroadcast` and `videobonusDetail` in `__NEXT__` data
* id must be `compat_str`
* return single video for singleton playlist.

Resolves #30151
Resolves #30166
Resolves #30172
Resolves #30406
Resolves #30696
Expected to resolve #29415.

### CODE REVIEW

File: youtube_dl/extractor/ceskatelevize.py [hunk 1]
```diff
@@ -12,70 +12,243 @@
...
     USER_AGENTS,
 )
 
+try:
+    from ..utils import traverse_obj
+except ImportError:
+    from ..compat import (compat_str, compat_collections_abc)
+    from ..utils import int_or_none
+
+    def traverse_obj(obj, *path_list, **kw):
```
Comment (pukkandan) [inline youtube_dl/extractor/ceskatelevize.py:26]: if u r going to port `traverse_obj`, port it properly to utils imho. It is very useful and superseeds most instances of `try_get` and `dict_get` while being able to do much more.

If for just this extractor, you can easily replace traverse_obj with equivalent `try_get`s

Comment (dirkf) [reply youtube_dl/extractor/ceskatelevize.py:26]: Thanks. I am going to port `traverse_obj()`, so as to make it easy to port extractors back and forth. This is actually my best current effort at a candidate for utils, but I had in mind at some point in the future to pull various things from yt-dlp/utils in a separate PR (that will delete this version). So this is a stopgap to get this extractor ported without changing the extractor code unnecessarily. Although extractors might not need all the features, I'd rather not have to change tests or disappoint someone who wanted to upgrade the simple `--print` proposed elsewhere.

File: youtube_dl/extractor/ceskatelevize.py [hunk 2]
```diff
@@ -12,70 +12,243 @@
...
+                        val = from_iterable(v for v in val if v is not None)
+                    val = [v for v in map(type_test, val) if v is not None]
+                    if val:
+                        return val if get_all else val[0]
+                else:
+                    val = type_test(val)
+                    if val is not None:
+                        return val
+        return default
+
```
Comment (dirkf) [inline youtube_dl/extractor/ceskatelevize.py:127]: No longer required.
```suggestion
```

File: youtube_dl/extractor/ceskatelevize.py [hunk 3]
```diff
@@ -12,70 +12,135 @@
     ExtractorError,
     float_or_none,
     sanitized_Request,
-    unescapeHTML,
-    update_url_query,
+    str_or_none,
```
Comment (dirkf) [inline youtube_dl/extractor/ceskatelevize.py:15]:
```suggestion
    str_or_none,
    traverse_obj,
```

**Pull request merged**